I embed window-tool-bar right into the active tab (at the top!). Only if the tab is active will it embed it. So the following can happen:

1. You click on a currently-inactive tab to the right of the currently-active tab.
2. The system redraws such that suddenly the currently-active tab needs a lot less space than before (since it's being deactivated) and the newly-active tab gets all the buttons.
3. The system things you clicked on one of the buttons when all you wanted is to activate a different tab
4. It does random shit.

In order to prevent that, react on mouse button press (which won't come in this situation--being already in the past) instead of on mouse button release (which WILL come shortly).

diff -ru orig/window-tool-bar-0.3/window-tool-bar.el window-tool-bar-0.3/window-tool-bar.el
--- orig/window-tool-bar-0.3/window-tool-bar.el	2025-02-08 23:02:18.000000000 +0100
+++ window-tool-bar-0.3/window-tool-bar.el	2025-02-10 00:04:52.645770356 +0100
@@ -185,21 +185,21 @@
   "<follow-link>" 'mouse-face
   ;; Follow link on all clicks of mouse-1 and mouse-2 since the tool
   ;; bar is not a place the point can travel to.
-  "<tab-line> <mouse-1>" #'window-tool-bar--call-button
-  "<tab-line> <double-mouse-1>" #'window-tool-bar--call-button
-  "<tab-line> <triple-mouse-1>" #'window-tool-bar--call-button
-  "<tab-line> <mouse-2>" #'window-tool-bar--call-button
-  "<tab-line> <double-mouse-2>" #'window-tool-bar--call-button
-  "<tab-line> <triple-mouse-2>" #'window-tool-bar--call-button
+  "<tab-line> <down-mouse-1>" #'window-tool-bar--call-button
+  "<tab-line> <double-down-mouse-1>" #'window-tool-bar--call-button
+  "<tab-line> <triple-down-mouse-1>" #'window-tool-bar--call-button
+  "<tab-line> <down-mouse-2>" #'window-tool-bar--call-button
+  "<tab-line> <double-down-mouse-2>" #'window-tool-bar--call-button
+  "<tab-line> <triple-down-mouse-2>" #'window-tool-bar--call-button
 
   ;; Mouse down events do nothing.  A binding is needed so isearch
   ;; does not exit when the tab bar is clicked.
-  "<tab-line> <down-mouse-1>" #'window-tool-bar--ignore
-  "<tab-line> <double-down-mouse-1>" #'window-tool-bar--ignore
-  "<tab-line> <triple-down-mouse-1>" #'window-tool-bar--ignore
-  "<tab-line> <down-mouse-2>" #'window-tool-bar--ignore
-  "<tab-line> <double-down-mouse-2>" #'window-tool-bar--ignore
-  "<tab-line> <triple-down-mouse-2>" #'window-tool-bar--ignore)
+  "<tab-line> <mouse-1>" #'window-tool-bar--ignore
+  "<tab-line> <double-mouse-1>" #'window-tool-bar--ignore
+  "<tab-line> <triple-mouse-1>" #'window-tool-bar--ignore
+  "<tab-line> <mouse-2>" #'window-tool-bar--ignore
+  "<tab-line> <double-mouse-2>" #'window-tool-bar--ignore
+  "<tab-line> <triple-mouse-2>" #'window-tool-bar--ignore)
 (fset 'window-tool-bar--button-keymap window-tool-bar--button-keymap) ; So it can be a keymap property
 
 ;; Register bindings that stay in isearch.  Technically, these
 diff -ru orig/window-tool-bar-0.3/window-tool-bar.el window-tool-bar-0.3/window-tool-bar.el
--- orig/window-tool-bar-0.3/window-tool-bar.el	2025-02-08 23:02:18.000000000 +0100
+++ window-tool-bar-0.3/window-tool-bar.el	2025-02-10 00:04:52.645770356 +0100
@@ -381,7 +381,11 @@ MENU-ITEM is a menu item to convert.  See info node `(elisp)Tool Bar'."
                                   help-text)
                                 str))
            (put-text-property 0 len 'tool-bar-key key str)
-           str))))))
+           str))))
+
+    ;; Non-menu items that don't get a button.
+    (`(,_ . ,(pred symbolp))
+     nil)))
 
 (defun window-tool-bar--call-button ()
   "Call the button that was clicked on in the tab line."
@@ -393,6 +397,8 @@ MENU-ITEM is a menu item to convert.  See info node `(elisp)Tool Bar'."
       (select-window (posn-window posn))
       (let* ((str (posn-string posn))
              (key (get-text-property (cdr str) 'tool-bar-key (car str)))
+             ;; FIXME: Use modifier keys which may have a different
+             ;; binding.
              (cmd (lookup-key (window-tool-bar--get-keymap) (vector key))))
         (call-interactively cmd)))))
 
